src/wayland/protocols/toplevel_info.rs

```diff
@@ -287,22 +348,32 @@ where
             } else {
                 for handle in &state.instances {
                     // don't send events to stopped instances
-                    if self
-                        .instances
-                        .iter()
-                        .any(|i| i.id().same_client_as(&handle.id()))
+                    if handle.version() < zcosmic_toplevel_info_v1::REQ_GET_COSMIC_TOPLEVEL_SINCE
+                        && self
+                            .instances
+                            .iter()
+                            .any(|i| i.id().same_client_as(&handle.id()))
                     {
                         handle.closed();
                     }
                 }
                 false
             }
         });
+        for instance in &self.instances {
+            if instance.version() >= zcosmic_toplevel_info_v1::EVT_DONE_SINCE {
+                instance.done();
```

Comment:
Testing this, I'm seeing a flood of `zcosmic_toplevel_info_v1::done` events.

Presumably it shouldn't send `done` unless at least one toplevel has changed (or been added or removed).

Reply:
Should be fixed now. Global `done` also needed some fixes to be only send after we cleaned up all of the state updates successfully.